module/os/linux/zfs/zfs_file_os.c

```diff
@@ -0,0 +1,312 @@
+{
+	filp_close(fp, 0);
+}
+
+int
+zfs_file_getattr(zfs_file_t *filp, zfs_file_attr_t *zfap)
+{
+	struct kstat stat;
+	int rc;
+
+	ASSERT(filp);
+	ASSERT(zfap);
+
+#if defined(HAVE_4ARGS_VFS_GETATTR)
+	rc = vfs_getattr(&filp->f_path, &stat, STATX_BASIC_STATS,
+	    AT_STATX_SYNC_AS_STAT);
+#elif defined(HAVE_2ARGS_VFS_GETATTR)
+	rc = vfs_getattr(&filp->f_path, &stat);
+#else
+	rc = vfs_getattr(filp->f_path.mnt, filp->f_dentry, &stat);
+#endif
+	if (rc)
+		return (-rc);
+
+
+	zfap->zfa_size = stat.size;
+	zfap->zfa_mode = stat.mode;
+
+	return (0);
+}
+
+int
+zfs_file_fsync(zfs_file_t *filp, int flags)
+{
+	int datasync = 0;
+	int error;
+	int fstrans;
+
+	ASSERT(filp);
+
+	if (flags & FDSYNC)
+		datasync = 1;
+
+	/*
+	 * May enter XFS which generates a warning when PF_FSTRANS is set.
+	 * To avoid this the flag is cleared over vfs_sync() and then reset.
+	 */
+	fstrans = __spl_pf_fstrans_check();
+	if (fstrans)
+		current->flags &= ~(__SPL_PF_FSTRANS);
+	error = -vfs_fsync(filp, datasync);
+	if (fstrans)
+		current->flags |= __SPL_PF_FSTRANS;
+
+	return (error);
+}
+
+int
+zfs_file_pwrite(zfs_file_t *fp, const void *buf, size_t count, loff_t *offp,
+    ssize_t *resid)
+{
+	ssize_t rc;
+
+#if defined(HAVE_KERNEL_READ_PPOS)
+	rc = kernel_write(fp, buf, count, offp);
+#else
+	mm_segment_t saved_fs;
+
+	saved_fs = get_fs();
+	set_fs(KERNEL_DS);
+
+	rc = vfs_write(fp, (__force const char __user __user *)buf,
+	    count, offp);
+
+	set_fs(saved_fs);
+#endif
+	if (rc < 0)
+		return (-rc);
+	if (resid)
+		*resid = count - rc;
+	return (0);
+}
+
+int
+zfs_file_write(zfs_file_t *fp, const void *buf, size_t count, ssize_t *resid,
+    int flags)
+{
+	zfs_file_attr_t zfa;
+	int rc;
+
+	if (flags & FAPPEND) {
+		if ((rc = zfs_file_getattr(fp, &zfa)))
+			return (rc);
+
+		fp->f_pos = zfa.zfa_size;
+	}
+	return (zfs_file_pwrite(fp, buf, count, &fp->f_pos, resid));
+}
+
+int
+zfs_file_pread(zfs_file_t *fp, void *buf, size_t count, loff_t *offp,
+    ssize_t *resid)
+{
+	ssize_t rc;
+
+#if defined(HAVE_KERNEL_READ_PPOS)
+	rc = kernel_read(fp, buf, count, offp);
+#else
+	mm_segment_t saved_fs;
+
+	saved_fs = get_fs();
+	set_fs(KERNEL_DS);
+
+	rc = vfs_read(fp, (void __user *)buf, count, offp);
+	set_fs(saved_fs);
+#endif
+	if (rc < 0)
+		return (-rc);
+	if (resid)
+		*resid = count - rc;
+	return (0);
+}
+
+int
+zfs_file_read(zfs_file_t *fp, void *buf, size_t count, ssize_t *resid)
+{
+	return (zfs_file_pread(fp, buf, count, &fp->f_pos, resid));
+}
+
+int
+zfs_file_seek(zfs_file_t *fp, loff_t *offp, int whence)
+{
+	loff_t rc;
+
+	if (*offp < 0 || *offp > MAXOFFSET_T)
+		return (EINVAL);
+	rc = vfs_llseek(fp, *offp, whence);
+
+	if (rc < 0)
+		return (-rc);
+	*offp = rc;
+	return (0);
+}
+
+loff_t
+zfs_file_off(zfs_file_t *fp)
+{
+	return (fp->f_pos);
+}
+
+int
+zfs_file_unlink(const char *path)
+{
+	return (EOPNOTSUPP);
+}
+
+int
+zfs_file_fallocate(zfs_file_t *fp, int mode, loff_t offset, loff_t len)
+{
+	int error = EOPNOTSUPP;
+#ifdef FALLOC_FL_PUNCH_HOLE
```

Comment:
Similar to above, we should be able to take the `FALLOC_FL_PUNCH_HOLE` path unconditionally now.